### PR TITLE
Step-11: Integrate supabase

### PR DESCRIPTION
**Pull request by sta-wtag**
https://user-images.githubusercontent.com/122765730/222076483-96329d6b-4687-4456-8a21-074eb85ac33e.mov

### Background
Add [supabase](https://supabase.nuxtjs.org/) to the project.

### Acceptance Criteria

- [x] Add supabase as Database to persist app state.

- [x] Make API calls using fetch from store Actions and set response to store state using Mutations.

### CODE REVIEW

File: helpers/utils.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+// const triggerToast = (type) => {
+//   swal($nuxt.$t('alert.message.error'), {
+//     buttons: false,
+//     className: 'swal-basic ' + type === SUCCESS ? 'success' : 'error',
+//     iconHtml: '<img src="https://picsum.photos/100/100">',
+//     timer: 3000,
+//   });
+// };
+
+// export default triggerToast;
```
Comment (sai-wtag) [inline helpers/utils.js]: Can we please remove this if not needed ?

File: package.json [hunk 2]
```diff
@@ -35,11 +35,12 @@
     "@nuxtjs/eslint-config": "^6.0.1",
     "@nuxtjs/eslint-module": "^3.0.2",
     "@nuxtjs/style-resources": "^1.2.1",
+    "@supabase/supabase-js": "^2.8.0",
     "@testing-library/vue": "^5.8.3",
     "@vue/test-utils": "^1.3.3",
     "@vue/vue2-jest": "^27.0.0-alpha.2",
     "babel-core": "^7.0.0-bridge.0",
-    "babel-jest": "^29.4.2",
+    "babel-jest": "^",
```
Comment (sai-wtag) [inline package.json]: Is it the best way to not to specify any version ?

File: store/todos.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
-import { uuid } from 'uuidv4';
+import { v4 as uuidv4 } from 'uuid';
+import Database from '../helpers/database';
```
Comment (sai-wtag) [inline store/todos.js]: Can we import from the relative path ? e.g. `@/helpers/database` ?

File: store/todos.js [hunk 4]
```diff
@@ -8,6 +9,9 @@ import {
   PER_PAGE,
 } from '../constants';
```
Comment (sai-wtag) [inline store/todos.js]: ditto

**Review by sai-wtag — COMMENTED** [review]
Screenshot 2023-03-01 at 3 14 39 PM
Add Card is not aligned with other cards.

File: helpers/utils.js (hunk 1 above)
Comment (sta-wtag) [reply helpers/utils.js]: resolved

File: package.json (hunk 2 above)
Comment (sta-wtag) [reply package.json]: resolved

**Review by sai-wtag — COMMENTED** [review]
A loader should add while adding a task.

https://user-images.githubusercontent.com/122719191/222096009-8ed5aa46-ece2-4ced-a538-48a46ce7753e.mov

**Comment by sta-wtag** [comment]
> Screenshot 2023-03-01 at 3 14 39 PM Add Card is not aligned with other cards.

resolved

**Comment by sai-wtag** [comment]
After creating `search` state should be clear and task state should be shifted to `all`

https://user-images.githubusercontent.com/122719191/222097310-21df8dd8-7a63-4fc8-972d-57de71c469ed.mov

**Comment by sai-wtag** [comment]
While shifting to i18n, localisation warning is showing.

https://user-images.githubusercontent.com/122719191/222098286-f4e151d7-23aa-4e32-a1cb-6583ea0a75dc.mov

**Comment by sai-wtag** [comment]
Can we add dropdown for filters in mobile menu ?

https://user-images.githubusercontent.com/122719191/222099313-b9f54976-1e2a-4baf-8ffa-3c0f17458b9a.mov

**Review by sai-wtag — CHANGES_REQUESTED** [review]
Some changes required.

**Comment by sta-wtag** [comment]
> Can we add dropdown for filters in mobile menu ?

resolved

**Comment by sta-wtag** [comment]
> Some changes required.

resolved

**Comment by sta-wtag** [comment]
> A loader should add while adding a task.

resolved

**Comment by sai-wtag** [comment]
Apu, I think loader area while creating a task should be on the adding card only not in the entire screen (according to figma design)

https://user-images.githubusercontent.com/122719191/222332228-c7672663-e301-42f9-af62-3ecb47d62e07.mov

Screenshot 2023-03-02 at 10 36 14 AM

**Review by sai-wtag — APPROVED** [review]
Locally tested. LGTM. Thanks

File: components/TaskCard.vue [hunk 5]
```diff
@@ -1,3 +1,10 @@
+Skip to content Search or jump to… Pull requests Issues Codespaces Marketplace
+Explore @sta-wtag sta-wtag / TodoApp Public Fork your own copy of
+sta-wtag/TodoApp Code Issues Pull requests 7 Actions Projects Wiki Security
+Insights Settings TodoApp/components/TaskCard.vue @sta-wtag sta-wtag add line
+clump to task description and see more button to view details Latest commit
+a885c7e yesterday History 1 contributor 272 lines (248 sloc) 6.42 KB
+
```
Comment (hhu-wtag) [inline components/TaskCard.vue]: Unnecessary text

File: components/TaskCard.vue [hunk 6]
```diff
@@ -83,14 +90,13 @@
...
-import LoadingIcon from '@/components/buttons/LoadingIcon.vue';
-import EditIcon from '@/assets/svg/Edit.svg';
-import TickIcon from '@/assets/svg/Tick.svg';
-import global from '@/mixins/global';
-
+import DeleteIcon from '../assets/svg/Delete.svg';
+import EditIcon from '../assets/svg/Edit.svg';
+import TickIcon from '../assets/svg/Tick.svg';
+import global from '../mixins/global';
+import LoadingIcon from './buttons/LoadingIcon.vue';
```
Comment (hhu-wtag) [inline components/TaskCard.vue]: Can we use relative path here?

File: components/TaskCard.vue [hunk 7]
```diff
@@ -83,14 +90,13 @@
...
 export default {
-  name: 'TaskCard',
+  // name: 'TaskCard',
```
Comment (hhu-wtag) [inline components/TaskCard.vue]: Remove if not needed

File: components/TaskContainer.vue [hunk 8]
```diff
@@ -1,108 +1,125 @@
...
-import NoTaskLogo from '@/assets/svg/noTask.svg';
-import LoadingIcon from '@/components/buttons/LoadingIcon.vue';
-import global from '@/mixins/global';
+import { SUCCESS, ERROR } from '../constants';
+import PlusIcon from '../assets/svg/plusIcon.svg';
+import NoTaskLogo from '../assets/svg/noTask.svg';
+import LoadingIcon from '../components/buttons/LoadingIcon.vue';
+import global from '../mixins/global';
+import DeleteIcon from './buttons/DeleteIcon.vue';
+import FilterComponent from './buttons/FilterComponent.vue';
```
Comment (hhu-wtag) [inline components/TaskContainer.vue]: ditto

File: mixins/global.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
 import swal from 'sweetalert';
 
-import { SUCCESS, ERROR, SWALCOLORS } from '@/constants.js';
+import { SUCCESS, SWALCOLORS } from '../constants.js';
```
Comment (hhu-wtag) [inline mixins/global.js]: Why are we moving back to static path system?

File: store/todos.js [hunk 10]
```diff
@@ -61,42 +65,71 @@ export const getters = {
...
+      }
+
+      state.filterOptions.forEach((element) => (element.status = false));
+      state.filterOptions[0].status = true;
+      state.activeFilterOption = state.filterOptions[0];
+
+      commit('filterTaskList');
+
+      return { success: true };
+    } catch (err) {}
```
Comment (hhu-wtag) [inline store/todos.js]: Can we throw an error here?

File: store/todos.js [hunk 11]
```diff
@@ -61,42 +65,71 @@ export const getters = {
...
-    });
+      if (error) {
+        return { success: false };
+      }
+
+      // store response to allTodos
+      commit('filterTaskList');
+
+      return { success: true };
+    } catch (err) {}
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

File: store/todos.js [hunk 12]
```diff
@@ -61,42 +65,71 @@ export const getters = {
...
+        return { success: false };
+      }
+
+      // store response to allTodos
+      commit('filterTaskList');
+
+      commit('setCompleteRequest', false);
+
+      return { success: true, data: todo };
+    } catch (err) {}
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

File: store/todos.js [hunk 13]
```diff
@@ -117,16 +150,26 @@ export const actions = {
...
+      if (error) {
+        return { success: false, data: todo };
+      }
+
+      // store response to allTodos
+      commit('filterTaskList');
+      commit('setCompleteRequest', false);
+
+      return { success: true, data: todo };
+    } catch (err) {}
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: components/TaskCard.vue (hunk 6 above)
Comment (sta-wtag) [reply components/TaskCard.vue]: Adding relative path throughs error on test.

File: store/todos.js (hunk 10 above)
Comment (sta-wtag) [reply store/todos.js]: resolved

File: store/todos.js (hunk 11 above)
Comment (sta-wtag) [reply store/todos.js]: resolved

File: store/todos.js (hunk 12 above)
Comment (sta-wtag) [reply store/todos.js]: resolved

File: store/todos.js (hunk 13 above)
Comment (sta-wtag) [reply store/todos.js]: resolved

File: components/TaskContainer.vue (hunk 8 above)
Comment (sta-wtag) [reply components/TaskContainer.vue]: Adding relative path throughs error on test.

File: mixins/global.js (hunk 9 above)
Comment (sta-wtag) [reply mixins/global.js]: Adding relative path throughs error on test.

File: components/TaskContainer.vue (hunk 8 above)
Comment (hhu-wtag) [reply components/TaskContainer.vue]: We can investigate the problem. 
A good start [here](https://github.com/facebook/jest/issues/3499)

File: components/TaskCard.vue (hunk 5 above)
Comment (sta-wtag) [reply components/TaskCard.vue]: resolved

File: components/TaskCard.vue (hunk 6 above)
Comment (sta-wtag) [reply components/TaskCard.vue]: resolved

File: components/TaskCard.vue (hunk 7 above)
Comment (sta-wtag) [reply components/TaskCard.vue]: resolved

File: components/TaskContainer.vue (hunk 8 above)
Comment (sta-wtag) [reply components/TaskContainer.vue]: resolved

File: mixins/global.js (hunk 9 above)
Comment (sta-wtag) [reply mixins/global.js]: resolved

File: components/TaskContainer.vue (hunk 8 above)
Comment (sta-wtag) [reply components/TaskContainer.vue]: resolved

File: mixins/global.js (hunk 9 above)
Comment (sta-wtag) [reply mixins/global.js]: resolved

File: components/TaskCard.vue (hunk 7 above)
Comment (sta-wtag) [reply components/TaskCard.vue]: resolved

File: components/TaskCard.vue (hunk 6 above)
Comment (sta-wtag) [reply components/TaskCard.vue]: resolved

**Review by hhu-wtag — CHANGES_REQUESTED** [review]
image

- [ ] The spinner is not spinning on it's own center position.

https://user-images.githubusercontent.com/98801376/222947564-dca47541-0674-42c9-9b7e-8b2e245a9cb1.mov

- [ ] We can switch to one column view in this dimension
Screen Shot 2023-03-05 at 1 26 56 PM

- [ ] Mixed locales

image

- [ ] While adding a new task, header, task add button and filters should be visible but disabled. The main spinner should be a bit bigger and also has the **not spinning in its center** problem.


https://user-images.githubusercontent.com/98801376/222947829-af5da5ff-1e48-42a9-8b69-98f4b25216bd.mov

File: components/TaskContainer.vue [hunk 14]
```diff
@@ -429,4 +434,7 @@ export default {
     padding-bottom: 20px;
   }
 }
+// .no-task {
+//   margin-top: 25vh;
+// }
```
Comment (hhu-wtag) [inline components/TaskContainer.vue]: Can we remove it if not needed?

File: tests/unit/Header.spec.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+  //       debounced: debounce(mock, 0),
+  //     });
+
+  //     const input = wrapper.find('[data-testid="searchInputField"]');
+
+  //     input.setValue('test');
+  //     await input.trigger('keyup');
+
+  //     expect(mock).toHaveBeenCalled();
+  //   });
```
Comment (hhu-wtag) [inline tests/unit/Header.spec.js]: ditto

File: tests/unit/TaskContainer.spec.js [hunk 16]
```diff
@@ -43,18 +57,81 @@ function wrapperFactory() {
...
+
+    const result = await addMock('task1');
+
+    expect(result).toBe(true);
+  });
+});
+
+describe('@/components/TaskContainer.vue', () => {
+  it('helper function getDuration returns valid string', async () => {
+    const addMock = jest.spyOn(helpers, 'getDuration');
```
Comment (hhu-wtag) [inline tests/unit/TaskContainer.spec.js]:
```suggestion
    const getDurationFnMock = jest.spyOn(helpers, 'getDuration');
```

File: tests/unit/TaskContainer.spec.js [hunk 17]
```diff
@@ -43,18 +57,81 @@ function wrapperFactory() {
...
+
+    await addButton.trigger('submit');
+
+    expect(add).toHaveBeenCalled();
+  });
+});
+
+describe('@/components/TaskContainer.vue', () => {
+  it('helper function checkForm returns true on valid input', async () => {
+    const addMock = jest.spyOn(helpers, 'checkForm');
```
Comment (hhu-wtag) [inline tests/unit/TaskContainer.spec.js]:
```suggestion
    const checkFormFnMock = jest.spyOn(helpers, 'checkForm');
```

File: tests/unit/TaskContainer.spec.js [hunk 18]
```diff
@@ -43,18 +57,81 @@ function wrapperFactory() {
...
+      showAddCard: true,
+      loading: false,
+      titleInputError: false,
+    });
+
+    const addButton = wrapper.find('[data-testid="add-button"]');
+    const inputFieldToAddTask = wrapper.find('[data-testid="taskTitle"]');
+
+    await inputFieldToAddTask.setValue('test');
+    const add = jest.spyOn(wrapper.vm, 'submitForm');
```
Comment (hhu-wtag) [inline tests/unit/TaskContainer.spec.js]:
```suggestion
    const submitFormFnMock = jest.spyOn(wrapper.vm, 'submitForm');
```

**Review by hhu-wtag — APPROVED** [review]
- [x] The spinner inside `<TaskCard />` is still not spinning around it's own center.
Hint: The extra div inside `<LoadingIcon />` is the culprit

File: tests/unit/TaskContainer.spec.js (hunk 17 above)
Comment (sta-wtag) [reply tests/unit/TaskContainer.spec.js]: resolved

File: tests/unit/TaskContainer.spec.js (hunk 16 above)
Comment (sta-wtag) [reply tests/unit/TaskContainer.spec.js]: resolved

File: components/TaskContainer.vue (hunk 14 above)
Comment (sta-wtag) [reply components/TaskContainer.vue]: removed

File: .env.example [hunk 19]
```diff
@@ -0,0 +1,2 @@
+SUPABASE_URL=
+SUPABASE_KEY=
```
Comment (bia-wtag) [inline .env.example]: missing newline

File: .gitignore [hunk 20]
```diff
@@ -55,7 +55,6 @@ typings/
 # Optional eslint cache
 .eslintcache
 
-.DS_Store
```
Comment (bia-wtag) [inline .gitignore:58]: why did we remove .DS_Store?

File: components/Header.vue [hunk 21]
```diff
@@ -56,8 +58,8 @@ export default {
   },
   computed: {
     ...mapGetters({
-      locales: 'lang/getLocals',
-      currentLocale: 'lang/getCurrentLocale',
+      locales: 'locals/getLocals',
+      currentLocale: 'locals/getCurrentLocale',
```
Comment (bia-wtag) [inline components/Header.vue]: nitpicking: should be `locales`

File: components/SplachScreen.vue [hunk 22]
```diff
@@ -0,0 +1,119 @@
+<template>
+  <div class="primary-background splash-screen">
+    <SplashIcon />
+    <span class="splash-text margin-left-7">TODOS</span>
```
Comment (bia-wtag) [inline components/SplachScreen.vue:4]: can we use translation?

File: mixins/global.js [hunk 23]
```diff
@@ -9,21 +18,49 @@ export default {
...
+      } else if (type === ERROR && feature === DELETE) {
+        return $nuxt.$t('alert.delete.message.error');
+      } else if (type === ERROR && feature === SEARCH) {
+        return $nuxt.$t('alert.search.message.error');
+      } else if (type === ERROR && feature === COMPLETE) {
+        return $nuxt.$t('alert.complete.message.error');
+      } else {
+        $nuxt.$t('alert.message.error');
+      }
+    },
```
Comment (bia-wtag) [inline mixins/global.js:55]: `switch-case` would look cleaner!

**Review by bia-wtag — APPROVED** [review]
LGTM functionality-wise. Todos for now,

- Fix failing tests and introduce more tests if needed as there are effectively 3 tests only.
- Review the comments.

other than that, works nicely, great job 🎉

File: .env.example (hunk 19 above)
Comment (sta-wtag) [reply .env.example]: resolved

File: .gitignore (hunk 20 above)
Comment (sta-wtag) [reply .gitignore:58]: resolved

File: components/Header.vue (hunk 21 above)
Comment (sta-wtag) [reply components/Header.vue]: resolved

File: mixins/global.js (hunk 23 above)
Comment (sta-wtag) [reply mixins/global.js:55]: converted to switch-case. Thanks !

File: components/SplachScreen.vue (hunk 22 above)
Comment (sta-wtag) [reply components/SplachScreen.vue:4]: resolved

File: tests/unit/Header.spec.js (hunk 15 above)
Comment (sta-wtag) [reply tests/unit/Header.spec.js]: resolved

**Pull request merged**